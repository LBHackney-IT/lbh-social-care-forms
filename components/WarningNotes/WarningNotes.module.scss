@import 'node_modules/govuk-frontend/govuk/base';

.container {
  margin: 0;
}

.note {
  margin-bottom: 2rem;
  @include govuk-media-query($from: tablet) {
    display: flex;
    justify-content: space-between;
    align-items: flex-start;
  }
}

.noteDetails {
  display: grid;
  grid-template-columns: 14rem 1fr;
  grid-row-gap: 0.5rem;
  margin-bottom: 1rem;

  dt {
    font-weight: bold;
  }

  dd,
  dt {
    margin-top: 0;
  }
}
