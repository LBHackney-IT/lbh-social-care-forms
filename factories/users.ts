import { Factory } from 'fishery';

import { User } from 'types';

export const userFactory = Factory.define<User>(() => ({
  name: 'foo',
  hasAdminPermissions: true,
  hasChildrenPermissions: true,
  hasAdultPermissions: true,
  hasDevPermissions: false,
  isAuditable: false,
  email: 'foo@bar.com',
  permissionFlag: 'A',
  isAuthorised: true,
  isInWorkflowsPilot: false,
  isInSafeguardingReviewing: false,
  isInPlacementManagementUnit: false,
}));

export const mockedOnlyAdultUser = userFactory.build({
  hasAdminPermissions: false,
  hasChildrenPermissions: false,
  hasDevPermissions: false,
});

export const mockedOnlyChildUser = userFactory.build({
  hasAdminPermissions: false,
  hasAdultPermissions: false,
  hasDevPermissions: false,
});

export const mockedAdminUser = userFactory.build({
  hasDevPermissions: false,
});

export const mockedUser = userFactory.build();

export const mockedUserInWorkflowsPilot = userFactory.build({
  isInWorkflowsPilot: true,
});
