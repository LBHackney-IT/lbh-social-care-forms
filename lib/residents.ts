import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, User } from 'types';
import { getAuditingParams } from '../utils/auditing';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends LegacyResident {
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(({ addressList, ...resident }: ResidentBE): LegacyResident => {
    const address = addressList?.find(
      ({ displayAddressFlag }) => displayAddressFlag === 'Y'
    );
    return {
      ...resident,
      address: address && {
        address: address.addressLine1,
        postcode: address.postCode,
        uprn: address.uprn,
      },
    };
  });

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  user: User
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params: getAuditingParams(user),
  });
  return data;
};

export const normalisePhoneInput = (
  formData: Resident
): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Resident
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (formData: Resident): Promise<Resident> => {
  await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return formData;
};
