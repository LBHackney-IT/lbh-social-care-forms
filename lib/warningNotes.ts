import axios from 'axios';

import { WarningNote } from 'types';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

export const getWarningNotesByResident = async (
  personId: number
): Promise<WarningNote[] | []> => {
  const { data }: { data: { warningNotes: WarningNote[] } } = await axios.get(
    `${ENDPOINT_API}/residents/${personId}/warningnotes`,
    {
      headers,
    }
  );
  return data.warningNotes.filter((note) => note.status === 'open');
};

export const getWarningNote = async (
  warningNoteId: number
): Promise<WarningNote | undefined> => {
  const { data } = await axios.get(
    `${ENDPOINT_API}/warningnotes/${warningNoteId}`,
    {
      headers,
    }
  );
  return data;
};

export const addWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.post(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};

export const updateWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.patch(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};
