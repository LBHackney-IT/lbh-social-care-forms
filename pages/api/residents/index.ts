import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';
import { apiHandler } from 'lib/apiHandler';
import { middleware as csrfMiddleware } from 'lib/csrfToken';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    res.status(StatusCodes.UNAUTHORIZED);
    return;
  }
  if (!user.isAuthorised) {
    res.status(StatusCodes.FORBIDDEN);
    return;
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Residents get error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Resident post error:',
          (error as AxiosError)?.response?.data
        );
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
