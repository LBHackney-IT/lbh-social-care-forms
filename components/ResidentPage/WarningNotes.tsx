import { WarningNote } from 'types';
import { useWarningNotes } from 'utils/api/warningNotes';
import Link from 'next/link';
import { canManageCases } from 'lib/permissions';
import { useAuth } from 'components/UserContext/UserContext';
import { useResident } from 'utils/api/residents';
import { useWorker } from 'utils/api/workers';
import { prettyWorkerName } from 'lib/formatters';
import { formatDate } from 'utils/date';
import s from './WarningNotes.module.scss';

interface WarningNoteProps {
  note: WarningNote;
  socialCareId: number;
}

const WarningNoteAnnouncement = ({ note, socialCareId }: WarningNoteProps) => {
  const { user } = useAuth();
  const { data: resident } = useResident(socialCareId);
  const { data: workers } = useWorker({
    email: note.createdBy,
  });
  const worker = workers?.[0];

  return (
    <section className="lbh-page-announcement lbh-page-announcement--warning">
      <h3 className="lbh-page-announcement__title lbh-heading-h5">
        Warning: <span className={s.warning}>{note.noteType}</span>
      </h3>
      <div className="lbh-page-announcement__content">
        <p className="lbh-body-xs">
          Added {formatDate(note.startDate)} by{' '}
          {worker && prettyWorkerName(worker)}
          {note.reviewDate && ` · Last reviewed ${formatDate(note.reviewDate)}`}
          {user && resident && canManageCases(user, resident) && (
            <>
              {' '}
              ·{' '}
              <Link href={`/people/${socialCareId}/warning-notes/${note.id}`}>
                <a className="govuk-link lbh-link">Review or end</a>
              </Link>
            </>
          )}
        </p>
      </div>
    </section>
  );
};

interface Props {
  socialCareId: number;
}

const WarningNotes = ({ socialCareId }: Props): React.ReactElement | null => {
  const { data } = useWarningNotes(socialCareId);

  if (data && data.length > 0)
    return (
      <>
        {data?.map((note) => (
          <WarningNoteAnnouncement
            note={note}
            key={note.id}
            socialCareId={socialCareId}
          />
        ))}
      </>
    );

  return null;
};

export default WarningNotes;
