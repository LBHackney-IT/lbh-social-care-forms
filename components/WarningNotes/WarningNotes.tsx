import cx from 'classnames';
import Link from 'next/link';

import { useWarningNotes } from 'utils/api/warningNotes';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { User, WarningNote } from 'types';

import styles from './WarningNotes.module.scss';
import { useResident } from '../../utils/api/residents';
import { canManageCases } from '../../lib/permissions';
import { useAuth } from '../UserContext/UserContext';

export interface Props {
  personId: number;
  notes: WarningNote[];
}

export const WarningBox = ({ notes, personId }: Props): React.ReactElement => {
  const { data: resident } = useResident(personId);
  const { user } = useAuth() as { user: User };

  if (!resident) {
    return <></>;
  }

  return (
    <div
      className={cx(
        'govuk-error-summary govuk-!-margin-bottom-8',
        styles.container
      )}
      aria-labelledby="warning-note-title"
      role="complementary"
      tabIndex={-1}
    >
      <h2 className="govuk-error-summary__title" id="warning-note-title">
        WARNING NOTE
      </h2>
      <div>
        <div className="govuk-error-summary__body">
          {notes.map((note) => (
            <div key={note.id} className={styles.note}>
              <dl className={styles.noteDetails}>
                {note.noteType && (
                  <>
                    <dt>Type</dt>
                    <dd>{note.noteType}</dd>
                  </>
                )}
                {note.startDate && (
                  <>
                    <dt>Start date</dt>
                    <dd>
                      {note.reviewedDate
                        ? new Date(note.reviewedDate).toLocaleDateString(
                            'en-GB'
                          )
                        : new Date(note.startDate).toLocaleDateString(
                            'en-GB'
                          )}{' '}
                      <i>created by</i> {note.createdBy}
                    </dd>
                  </>
                )}
                {note.reviewDate && !note.nextReviewDate && (
                  <>
                    <dt>Review date</dt>
                    <dd>
                      {new Date(note.reviewDate).toLocaleDateString('en-GB')}{' '}
                    </dd>
                  </>
                )}
                {note.nextReviewDate && (
                  <>
                    <dt>Next review date</dt>
                    <dd>
                      {new Date(note.nextReviewDate).toLocaleDateString(
                        'en-GB'
                      )}{' '}
                    </dd>
                  </>
                )}
              </dl>
              {canManageCases(user, resident) && (
                <Link href={`/people/${personId}/warning-notes/${note.id}`}>
                  <a className="govuk-link govuk-link-underline">
                    Review / end
                  </a>
                </Link>
              )}
            </div>
          ))}
          <p>For further information see Warning Note in Records History</p>
        </div>
      </div>
    </div>
  );
};

const WarningNotes = ({ id }: { id: number }): React.ReactElement | null => {
  const { data: warningNotes, error } = useWarningNotes(id);
  if (error) {
    return <ErrorMessage />;
  }
  if (!warningNotes || warningNotes.length === 0) {
    return null;
  }
  return <WarningBox notes={warningNotes} personId={id} />;
};

export default WarningNotes;
