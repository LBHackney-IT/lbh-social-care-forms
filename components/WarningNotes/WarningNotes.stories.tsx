import { WarningBox, Props } from './WarningNotes';
import { Story } from '@storybook/react';

import { mockedWarningNote } from 'factories/warningNotes';

export default {
  title: 'Warning Note',
  component: WarningBox,
};

const Template: Story<Props> = (args) => <WarningBox {...args} />;

export const Default = Template.bind({});
Default.args = {
  personId: 123,
  notes: mockedWarningNote.slice(0, 1),
};

export const multiWarningNote = Template.bind({});
multiWarningNote.args = {
  personId: 123,
  notes: mockedWarningNote,
};
