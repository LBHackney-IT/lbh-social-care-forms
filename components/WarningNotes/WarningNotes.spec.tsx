import { render } from '@testing-library/react';

import * as residentsAPI from 'utils/api/residents';
import * as warningNotes from 'utils/api/warningNotes';
import WarningNotes from './WarningNotes';

import { warningNoteFactory } from 'factories/warningNotes';
import { AuthProvider } from '../UserContext/UserContext';
import { userFactory } from '../../factories/users';
import { residentFactory } from '../../factories/residents';

const mockedAdultsUser = userFactory.build({
  hasAdminPermissions: false,
  hasUnrestrictedPermissions: false,
  hasChildrenPermissions: false,
  hasAdultPermissions: true,
});

const mockedAdultsResident = residentFactory.build({
  contextFlag: 'A',
});

jest.mock('next/router', () => ({
  useRouter: () => ({
    asPath: 'path',
    push: jest.fn(),
  }),
}));

describe(`useWarningNotes`, () => {
  beforeEach(() => {
    jest.spyOn(warningNotes, 'useWarningNotes').mockImplementation(() => ({
      data: [
        warningNoteFactory.build({
          nextReviewDate: '2020-11-12',
        }),
      ],
      mutate: jest.fn(),
      revalidate: jest.fn(),
      isValidating: false,
    }));

    jest.spyOn(residentsAPI, 'useResident').mockImplementation(() => ({
      data: mockedAdultsResident,
      isValidating: false,
      mutate: jest.fn(),
      revalidate: jest.fn(),
    }));
  });

  it('should render a list of warning notes', () => {
    const { asFragment } = render(
      <AuthProvider user={mockedAdultsUser}>
        <WarningNotes id={123} />
      </AuthProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not render the review date if there is a scheduled future review date', () => {
    const { getByText, queryByText } = render(
      <AuthProvider user={mockedAdultsUser}>
        <WarningNotes id={123} />
      </AuthProvider>
    );

    getByText('Next review date');

    expect(
      queryByText('Review date', {
        exact: true,
      })
    ).toBeNull();
  });
});
