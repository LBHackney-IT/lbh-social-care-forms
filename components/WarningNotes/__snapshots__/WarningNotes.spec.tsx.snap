// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWarningNotes should render a list of warning notes 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="warning-note-title"
    class="govuk-error-summary govuk-!-margin-bottom-8 container"
    role="complementary"
    tabindex="-1"
  >
    <h2
      class="govuk-error-summary__title"
      id="warning-note-title"
    >
      WARNING NOTE
    </h2>
    <div>
      <div
        class="govuk-error-summary__body"
      >
        <div
          class="note"
        >
          <dl
            class="noteDetails"
          >
            <dt>
              Type
            </dt>
            <dd>
              Risk to Staff
            </dd>
            <dt>
              Start date
            </dt>
            <dd>
              22/11/2020 
              <i>
                created by
              </i>
               Foo
            </dd>
            <dt>
              Next review date
            </dt>
            <dd>
              12/11/2020 
            </dd>
          </dl>
          <a
            class="govuk-link govuk-link-underline"
            href="/people/123/warning-notes/3"
          >
            Review / end
          </a>
        </div>
        <p>
          For further information see Warning Note in Records History
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
